django/db/models/expressions.py

```diff
@@ -1461,14 +1458,17 @@ def as_sql(
         template_params = {**self.extra, **extra_context}
         case_parts = []
         sql_params = []
+        default_sql, default_params = compiler.compile(self.default)
         for case in self.cases:
             try:
                 case_sql, case_params = compiler.compile(case)
             except EmptyResultSet:
                 continue
+            except FullResultSet:
+                default_sql, default_params = compiler.compile(case.result)
+                break
```

Comment:
As soon as a case is confirmed to always match, default to its results as other cases are unreachable. Think of it as branch pruning for `if x: ...; elif True: ...; elif y: ...; else ...``; the `elif y: ...; else ...` branch is unreachable.

This optimization paired with the `if not case parts` below makes that a `CASE WHEN full_result_set THEN "foo"` get optimized to `"foo"`.